### ISSUE TITLE
Multithreading Question

### ISSUE THREAD

**Issue by laxnpander** [LiangliangNan/Easy3D#110]
Hey hey,

it's me again. I am back at my original work and kind of want to resolve my visualisation issues once and for all. It's rather a question than an issue, so feel free to close it as you please.

My question is how to handle multithreading if I don't care about performance problems? I attached a naive approach how I would want to use easy3d but can't, because it throws exceptions. 

What would be the correct way to do it? I remember you noted that GL contexts live in the main thread, so I have to update from the main thread. But then again - in the main thread lives my viewer.run(). So how do I modify / extend / clear my point cloud, after the viewer was already started?

Thanks for the answer in advance!

```c
#include <easy3d/viewer/viewer.h>
#include <easy3d/renderer/camera.h>
#include <easy3d/renderer/drawable_lines.h>
#include <easy3d/renderer/drawable_points.h>
#include <easy3d/renderer/drawable_triangles.h>
#include <easy3d/fileio/resources.h>
#include <easy3d/core/types.h>
#include <easy3d/util/logging.h>
#include <easy3d/core/point_cloud.h>
#include <easy3d/renderer/drawable_points.h>
#include <easy3d/renderer/drawable_lines.h>
#include <easy3d/renderer/renderer.h>

#include <thread>


using namespace easy3d;

void addPoints(Viewer* viewer)
{
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    viewer->clear_scene();
    
    auto cloud = new PointCloud;
    cloud->add_vertex(easy3d::vec3(1.0, 1.0, 1.0));
    cloud->add_vertex(easy3d::vec3(4.0, 1.0, 1.0));
    cloud->add_vertex(easy3d::vec3(4.0, 4.0, 1.0));
    cloud->add_vertex(easy3d::vec3(1.0, 4.0, 1.0));
    cloud->add_vertex(easy3d::vec3(1.0, 1.0, 2.0));
    cloud->add_vertex(easy3d::vec3(4.0, 1.0, 2.0));
    cloud->add_vertex(easy3d::vec3(4.0, 4.0, 2.0));
    cloud->add_vertex(easy3d::vec3(1.0, 4.0, 2.0));
    viewer->add_model(cloud);
    
    viewer->update();
    std::cout << "Finished." << std::endl;
}

int main(int argc, char **argv) {

    logging::initialize();

    Viewer viewer("Tutorial_100_Test");

    auto cloud = new PointCloud;
    
    cloud->add_vertex(easy3d::vec3(1.0, 1.0, 1.0));
    cloud->add_vertex(easy3d::vec3(4.0, 1.0, 1.0));
    cloud->add_vertex(easy3d::vec3(4.0, 4.0, 1.0));
    cloud->add_vertex(easy3d::vec3(1.0, 4.0, 1.0));
    
    viewer.add_model(cloud);
    
    std::thread t(addPoints, &viewer);
    t.detach();
    
    viewer.fit_screen();

    return viewer.run();
}


```

**Comment by LiangliangNan** [comment]
OpenGL function calls can only appear in the main thread (the one in which the OpenGL context was created). So the viewer's `clear_scene()` and `add_model()` functions can NOT be in that thread, because they both involve OpenGL calls.

The following will work:
  - you can do whatever editing to the point cloud (e.g., adding points, deleting points from a point cloud) in any thread;
  - after editing the point cloud, call `cloud->renderer()->update()` to notify the viewer to update the visualization. The `cloud->renderer()->update()` itself doesn't call any OpenGL functions.

Try the code below:
-----------------------------

#include <easy3d/viewer/viewer.h>
#include <easy3d/renderer/camera.h>
#include <easy3d/renderer/drawable_lines.h>
#include <easy3d/renderer/drawable_points.h>
#include <easy3d/renderer/drawable_triangles.h>
#include <easy3d/fileio/resources.h>
#include <easy3d/core/types.h>
#include <easy3d/util/logging.h>
#include <easy3d/core/point_cloud.h>
#include <easy3d/renderer/drawable_points.h>
#include <easy3d/renderer/drawable_lines.h>
#include <easy3d/renderer/renderer.h>

#include <thread>


using namespace easy3d;


// Drawables are typically for rendering 3D models (e.g., point clouds, meshes, graphs) that are loaded from files or
// generated by some algorithms.
// The use of drawables for visualization is quite flexible. Drawables are normally attached to a 3D model. For example,
// you can attach a TrianglesDrawable to a surface mesh to visualize its surface and a PointsDrawable to visualize its
// vertices.
// Easy3D also allows visualizing stand-alone drawables (without creating a model).
//
// This example shows how to
//      - visualize 3D data without explicitly defining a model (i.e., rendering drawables directly);
//          * a set of triangles;
//          * a set of points;
//          * a set of lines.
//		- create a drawable for a specific rendering purpose;
//		- use the viewer to visualize drawables.

void addPoints(PointCloud* cloud)
{
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    for (int i=0; i<100; ++i)
        cloud->add_vertex(easy3d::vec3(random_float(), random_float(), random_float()));

    cloud->renderer()->update();
}


int main(int argc, char **argv) {
    srand (time(NULL));

    // Initialize logging.
    logging::initialize();

    Viewer viewer("Tutorial_100_Test");

    auto cloud = new PointCloud;
    cloud->add_vertex(easy3d::vec3(random_float(), random_float(), random_float()));
    cloud->add_vertex(easy3d::vec3(random_float(), random_float(), random_float()));
    cloud->add_vertex(easy3d::vec3(random_float(), random_float(), random_float()));
    cloud->add_vertex(easy3d::vec3(random_float(), random_float(), random_float()));
    viewer.add_model(cloud);

    auto drawable = cloud->renderer()->get_points_drawable("vertices");
    drawable->set_point_size(13.0f);

    std::thread t(addPoints, cloud);
    t.detach();

    viewer.fit_screen();

    // Run the viewer
    return viewer.run();
}

**Comment by laxnpander** [comment]
```c
cloud->renderer()->update();
```
This is it! This is my missing puzzle piece! That was exactly what I needed, again thank you very much! 
Small follow up question: Updating color / position of points already in the cloud is possible? The following example obviously crashes again due to update_vertex_buffer and update_color_buffer. But without calling it, color information is not being updated.

```c
#include <easy3d/viewer/viewer.h>
#include <easy3d/renderer/camera.h>
#include <easy3d/renderer/drawable_lines.h>
#include <easy3d/renderer/drawable_points.h>
#include <easy3d/renderer/drawable_triangles.h>
#include <easy3d/fileio/resources.h>
#include <easy3d/core/types.h>
#include <easy3d/util/logging.h>
#include <easy3d/core/point_cloud.h>
#include <easy3d/renderer/drawable_points.h>
#include <easy3d/renderer/drawable_lines.h>
#include <easy3d/renderer/renderer.h>

#include <thread>

using namespace easy3d;

void addPoints(PointCloud* cloud)
{
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    
    auto points = cloud->get_vertex_property<easy3d::vec3>("v:point");
    auto colors = cloud->add_vertex_property<easy3d::vec3>("v:color");
    
    for (auto v : cloud->vertices())
    	colors[v] = easy3d::vec3(1.0, 0.0, 0.0);
    	
    // Bad human begin
    auto drawable = cloud->renderer()->get_points_drawable("vertices");
    drawable->update_vertex_buffer(points.vector());
    drawable->update_color_buffer(colors.vector());
    // Bad human end
    
    for (int i=0; i<100; ++i)
        cloud->add_vertex(easy3d::vec3(random_float(), random_float(), random_float()));

    cloud->renderer()->update();
}


int main(int argc, char **argv) {
    srand (time(NULL));

    // Initialize logging.
    logging::initialize();

    Viewer viewer("Tutorial_100_Test");

    auto cloud = new PointCloud;
    cloud->add_vertex(easy3d::vec3(random_float(), random_float(), random_float()));
    cloud->add_vertex(easy3d::vec3(random_float(), random_float(), random_float()));
    cloud->add_vertex(easy3d::vec3(random_float(), random_float(), random_float()));
    cloud->add_vertex(easy3d::vec3(random_float(), random_float(), random_float()));
    viewer.add_model(cloud);

    auto drawable = cloud->renderer()->get_points_drawable("vertices");
    drawable->set_point_size(13.0f);

    std::thread t(addPoints, cloud);
    t.detach();

    viewer.fit_screen();

    // Run the viewer
    return viewer.run();
}
```

**Comment by LiangliangNan** [comment]
You need to set up the rendering in the main thread. And in the worker thread, you only modify the model but not call any OpenGL-related functions.

I've modified your code and made this officially part of the easy3d example. 
See:
https://github.com/LiangliangNan/Easy3D/blob/main/tutorials/Tutorial_312_MultiThread/main.cpp

**Comment by laxnpander** [comment]
@LiangliangNan Ah, once again this makes perfect sense and I feel stupid for not having thought about it that way. Thanks for pointing this out correctly, it looks like everything is running smooth now! Amazing support, crazy response time. This repo is about to go to the moon :-)

**Comment by laxnpander** [comment]
@LiangliangNan hmm, okay I think I understand. So for example if I want to render a trajectory (as a series of lines) that is being extended every second, I need to implement my own „TrajectoryDrawable“ and provide said update function, which describes how to fill the GL buffers?

**Comment by LiangliangNan** [comment]
You don't have to implement any drawable. The existing `LinesDrawable` is what you need, and you only need to implement the update function.

You can run `Tutorial_204_CameraInterpolation` to see what the existing camera trajectories look like.

**Comment by LiangliangNan** [comment]
It looks like this:

![Screen Shot 2022-02-18 at 23 12 20](https://user-images.githubusercontent.com/15526536/154768201-55c4a35c-395e-45b9-a694-ba00006ac56e.png)

**Comment by laxnpander** [comment]
Ah, got it! Thanks for your patience! I will investigate the next few days and reopen if I encounter any problem!

**Comment by LiangliangNan** [comment]
@laxnpander  I added some example code to show how to

* use the built-in drawables of Easy3D and provide customized update functions:
https://github.com/LiangliangNan/Easy3D/blob/89525eb8cbadf38fca2f182926204e959f282493/tutorials/Tutorial_301_Drawables/main.cpp#L152 

* inherit customized drawables from Easy3D drawables and  reimplement "void update_buffers_internal()":
https://github.com/LiangliangNan/Easy3D/blob/89525eb8cbadf38fca2f182926204e959f282493/tutorials/Tutorial_301_Drawables/main.cpp#L259

**Comment by laxnpander** [comment]
@LiangliangNan Nice, very helpful! Is it sufficient to set the update_func? Or do i need to request the drawable to update via the drawable->update()? My code looks something like this now:
```c
  // Creating the trajectory drawable
  auto m_draw_traj = new easy3d::LinesDrawable("trajectory");
  m_draw_traj->set_model(m_viewer->current_model());
  m_draw_traj->set_uniform_coloring(easy3d::vec4(0.0f, 0.0f, 1.0f, 1.0f));    // r, g, b, a
  m_draw_traj->set_line_width(1.0f);
  m_draw_traj->set_update_func([this](easy3d::Model* m, easy3d::Drawable* d) 
  {
    d->update_vertex_buffer(m_traj_points);        // m_traj_points is a member holding the trajectory translations
    d->update_element_buffer(m_traj_indices);   // m_traj_indices is a member holding the relations between connected frames
  });
  m_viewer->add_drawable(m_draw_traj);
```
Then in another function I basically modify m_traj_points and m_traj_indices to update the trajectory. However, nothing seems to happen. I added a print statement to the my update_func lambda and it is called exactly once. But when I add a call to
```c
m_draw_traj->update();
```
the code crashes with 
```
*** Check failure stack trace (most recent call first): ***
	1  	libspear            	easy3d::GenericBox<3, float>::clear()
	2  	libspear            	easy3d::Drawable::update()
	3  	libspear            	spear::Easy3d::renderTrajectory()
	4  	libspear            	spear::Easy3d::update()
	7  	libspear            	spear::WorkerThread::run()
	10 	libc.so             	clone

```

**Comment by LiangliangNan** [comment]
can you post the entire code or a minimum code that reproduces the crash？

**Comment by laxnpander** [comment]
@LiangliangNan I found the problem, it was a mistake of mine. My m_draw_traj was declared as local variable and shadowed the member. Now it's working flawlessly! Thank you very much! Btw. only using this line
```c
m_draw_traj->set_model(m_viewer->current_model());
```
was not sufficient for the update of the drawable. I had to add it to the viewer with
```c
m_viewer->add_drawable(m_draw_traj);
```
See also full code attached.

```c
#include <easy3d/viewer/viewer.h>
#include <easy3d/renderer/camera.h>
#include <easy3d/renderer/drawable_lines.h>
#include <easy3d/renderer/drawable_points.h>
#include <easy3d/renderer/drawable_triangles.h>
#include <easy3d/fileio/resources.h>
#include <easy3d/core/types.h>
#include <easy3d/util/logging.h>
#include <easy3d/core/point_cloud.h>
#include <easy3d/renderer/drawable_points.h>
#include <easy3d/renderer/drawable_lines.h>
#include <easy3d/renderer/renderer.h>

#include <thread>


using namespace easy3d;

std::vector<vec3> g_points;
std::vector<unsigned int> g_indices;

void addPoints(Viewer* viewer, const std::vector<easy3d::vec3> &pts)
{
    std::this_thread::sleep_for(std::chrono::milliseconds(300));
    
    auto lines = new easy3d::LinesDrawable("lines");
    //lines->set_model(viewer->current_model());
    lines->set_uniform_coloring(easy3d::vec4(0.0f, 0.0f, 1.0f, 1.0f));    // r, g, b, a
    lines->set_line_width(5.0f);
    lines->set_update_func([=](Model* m, Drawable* d)
    {
    	d->update_vertex_buffer(g_points, true);
    	d->update_element_buffer(g_indices);
    	std::cout << "Updating lines: " << g_points.size() << " points, " << g_indices.size() << " indices" << std::endl;
    });
    viewer->add_drawable(lines); // does not work without this line
    
    for (int i = 0; i < pts.size(); ++i)
    {
    	g_points.push_back(pts[i]);
    	if (i > 0)
    	{
    	   g_indices.push_back(i-1);
    	   g_indices.push_back(i);
        }
        lines->update();
    	viewer->update();
    	std::this_thread::sleep_for(std::chrono::milliseconds(100));
    	std::cout << "Iteration #" << i+1 << std::endl;
    }

    std::cout << "Finished." << std::endl;
}

int main(int argc, char **argv) {

    logging::initialize();

    Viewer viewer("Tutorial_100_Test");

    auto cloud = new PointCloud;
    
    std::vector<easy3d::vec3> pts{
    	easy3d::vec3(4.0, 1.0, 1.0),
    	easy3d::vec3(4.0, 4.0, 1.0),
    	easy3d::vec3(1.0, 4.0, 1.0),
    	easy3d::vec3(1.0, 1.0, 1.0),
    	easy3d::vec3(4.0, 1.0, 4.0),
    	easy3d::vec3(4.0, 4.0, 4.0),
    	easy3d::vec3(1.0, 4.0, 4.0),
    	easy3d::vec3(1.0, 1.0, 4.0),
    };
    
    for (const auto& pt : pts)
    	cloud->add_vertex(pt);
    
    viewer.add_model(cloud);
    
    auto drawable = cloud->renderer()->get_points_drawable("vertices");
    drawable->set_point_size(13.0f);
    
    
    std::thread t(addPoints, &viewer, pts);
    t.detach();
    
    viewer.fit_screen();

    return viewer.run();
}
```

**Comment by LiangliangNan** [comment]
> ```c
> ```c
> m_draw_traj->set_model(m_viewer->current_model());
> ```
> was not sufficient for the update of the drawable. I had to add it to the viewer with
> ```

Indeed, you're right. I misinterpreted it as `cloud->add_lines_drawable(...)`.

Glad it works. Good luck!
